LibWeb/DOM: Implement the Document object's fgColor, linkColor, vlinkColor, alinkColor, and bgColor attributes.

PGHales opened the pull request:
Implements document.bgColor in the Ladybird browser.

See issue [here](https://github.com/SerenityOS/serenity/issues/23929)

Edit: Updated to also implement fgColor, linkcolor, vlinkColor, and alinkColor since they were very similar.

awesomekling commented on Userland/Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -2289,6 +2289,22 @@ void Document::set_cookie(StringView cookie_string, Cookie::Source source)
 
+String Document::bg_color() const
+{
+    if (auto* body_element = body(); body_element && body_element->layout_node()) {
+        Color color = body_element->layout_node()->computed_values().background_color();
+        return color.to_string_without_alpha();
+    }
+    return ""_string;
```
If I understand correctly, other browsers simply return the `<body bgcolor>` attribute value verbatim here, i.e something like:

```cpp
String Document::bg_color() const
{
    if (auto* body_element = body())
        return body_element->get_attribute_value(HTML::AttributeNames::bgcolor);
    return ""_string;
}

void Document::set_bg_color(String const& value)
{
    if (auto* body_element = body())
        body_element->set_attribute(HTML::AttributeNames::bgcolor, value);
}
```

awesomekling commented:
From CI:
```
Files with no newline at the end: Tests/LibWeb/Text/input/DOM/Document-set-bgColor.html
```
Please configure your editor to put a newline at the end of files (Unix style)

awesomekling commented on Userland/Libraries/LibWeb/DOM/Document.idl:
```diff
@@ -57,6 +57,8 @@ interface Document : Node {
 
     attribute DOMString cookie;
 
+    attribute DOMString bgColor;
```
These should have `[CEReactions, LegacyNullToEmptyString]` per the spec: https://html.spec.whatwg.org/#Document-partial

awesomekling replied:
Also while you're here, you could also implement `fgColor`, `linkColor`, `vlinkColor`, `alinkColor` in exactly the same way as `bgColor` :slightly_smiling_face:

PGHales commented:
@awesomekling I went ahead and updated the PR with your suggestions. I implemented fgColor, linkColor, alinkColor, and vlinkColor in addition to bgColor that we already implemented since they were similar as you mentioned.

LucasChollet commented on Userland/Libraries/LibWeb/DOM/Document.h:
```diff
@@ -198,8 +213,8 @@ class Document
 
-    Color link_color() const;
-    void set_link_color(Color);
+    Color normal_link_color() const;
```
I am no maintainer and I know very little about `LibWeb`, but in general it's nice to keep commits atomic.
Here that would be, first renaming `link_color` in one commit and then add your changes to a second one.

tcl3 commented on Tests/LibWeb/Text/input/DOM/Document-set-alinkColor.html:
```diff
@@ -0,0 +1,13 @@
+<body>
```
As currently written, your test files don't run correctly. They should look something like this:
```html
<!DOCTYPE html>
<script src="../include.js"></script>
<script>
    test(() => {
        println("before: " + document.bgcolor);
        println("body bgcolor: " + document.body.getAttribute("bgcolor"));    
        
        document.bgcolor = "red";
    
        println("after:" + document.bgcolor);
        println("body bgcolor:" + document.body.getAttribute("bgcolor"));
    });
</script>
```
All of these tests look pretty similar. It might be worth seeing if you can combine them into a single test using a helper function.

---

FYI, you can run the LibWeb tests locally to ensure they run as expected with the following command: 
```sh
./Meta/serenity.sh test lagom LibWeb
```
You should also be able to view the test pages in Ladybird to verify that they look correct.

PGHales replied:
The tests are now combined into one test, but I can't verify if the tests run. When I run the command to run the tests, it says that "no tests were found". I'll try running it on my other computer setup later and see if it works there, but hopefully it is working now after the new changes.

tcl3 replied:
That's strange, it *should* just work. Does running `./Meta/serenity.sh run lagom ladybird` first make any difference?

Feel free to ask for help on Discord if you get stuck with this.

PGHales replied:
I got the tests working, I just redownloaded and built everything from scratch and that did the trick.

tcl3 commented on Userland/Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -2289,6 +2289,71 @@ void Document::set_cookie(StringView cookie_string, Cookie::Source source)
 
+String Document::fg_color() const
+{
+    if (auto* body_element = body())
```
The specification says that:

> [...] When there is no body element or if it is a frameset element, the attributes must instead return the empty string on getting and do nothing on setting.

So I think this something like this is needed in each of your getters and setters:

```suggestion
    if (auto* body_element = body(); body_element && !is<HTML::HTMLFrameSetElement>(*body_element))
```

This accounts for the case where `body()` returns a `HTMLFrameSetElement`.

---

FYI, this page has a test case that covers the behavior that is expected if `body` is a frameset element: 

http://wpt.live/html/obsolete/requirements-for-implementations/other-elements-attributes-and-apis/document-color-01.html

All of the tests on this page should pass if these attributes are implemented correctly.

tcl3 commented:
I've just seen that this has conflicts. please rebase on master and fix them up. :^)

tcl3 commented on Userland/Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -2285,6 +2285,71 @@ void Document::set_cookie(StringView cookie_string, Cookie::Source source)
+void Document::set_fg_color(String const& value)
+{
+    if (auto* body_element = body(); body_element && !is<HTML::HTMLFrameSetElement>(*body_element))
+        MUST(body_element->set_attribute(HTML::AttributeNames::text, value));
```
`TRY()` should be used here (and in the other setters), so that any error returned by `set_attribute()` is propagared to the caller instead of the program crashing. To do this you need to change the setters' return type to `WebIDL::ExceptionOr<void>`.

Sorry for the back and forth, I was sure I mentioned this in my last review :sweat_smile:

awesomekling replied:
I believe it's okay to use `MUST()` here as the only possible exception from `set_attribute()` is when you're passing an invalid attribute name.

tcl3 replied:
> I believe it's okay to use `MUST()` here as the only possible exception from `set_attribute()` is when you're passing an invalid attribute name.

You're absolutely right.

Ignore me, what's there currently is fine!

PGHales commented:
I think that should be everything, I got the new tests passing. Let me know if there is anything that still needs to be changed.

awesomekling commented:
Looks all right, nicely done :)
